### PR TITLE
[fix](Nereids) local aggregate output type is mismatch with BE

### PR DESCRIPTION
**Pull request by morrySnow**
## Problem summary

After aggregate disassemble, we break aggregate into two stage: LOCAL and GLOBAL.
All `AggregateFunction` in both of them use `getDataType` to get the output data type of the function.
But in BE, aggregate function has different output type according the stage it is belong to.
This PR fix this mismatch by:
- introduce two interface on `AggregateFunction`: `getLocalDataType` and `getGlobalDataType`
- introduce a new attribute in `AggregateFunction`: `isLocal`
- set `getDataType` as final on `AggregateFunction`
- `AggregateFunction#getDataType` choose `getLocalDataType` or `getGlobalDataType` according to `isLocal`

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/analysis/FunctionCallExpr.java [hunk 1]
```diff
@@ -1431,7 +1431,7 @@ public void finalizeImplForNereids() throws AnalysisException {
         // TODO: Supports type conversion to match the type of the function's parameters
         if (fnName.getFunction().equalsIgnoreCase("sum")) {
             // Prevent the cast type in vector exec engine
-            Type childType = getChild(0).type.getMaxResolutionType();
+            Type childType = getChild(0).type;
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/FunctionCallExpr.java:1434]: Why change this?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/AggregateFunction.java [hunk 2]
```diff
@@ -17,31 +17,59 @@
...
+    public AggregateFunction(String name, boolean isLocal, Expression... arguments) {
         super(name, arguments);
+        this.isLocal = isLocal;
+    }
+
+    public boolean isLocal() {
+        return isLocal;
     }
 
     public abstract DataType getIntermediateType();
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/AggregateFunction.java:46]: remove this?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/AggregateFunction.java [hunk 3]
```diff
@@ -17,31 +17,59 @@
...
 
+    /**
+     * When do two stage aggregate, the local stage's output type is different from the global.
+     * This method is used to get the local stage's output type.
+     */
+    public abstract DataType getLocalDataType();
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/AggregateFunction.java:52]: Why don't return the getDataType refer to the isLocal property, but use two distinct funciton to do this?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/AggregateFunction.java (hunk 2 above)
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/AggregateFunction.java:46]: this is useful in future

**Review by adonis0147 — APPROVED** [review]
LGTM

**Pull request closed without merging**